File: 2300-successful-pairs-of-spells-and-potions-1.java
import java.util.*;

class SuccessfulPairsOfSpellsAndPotionsSolution1 {

    public int[] successfulPairs(int[] spells, int[] potions, long success) {
        int n = spells.length, m = potions.length;
        Arrays.sort(potions);
        int[] result = new int[n];
        for (int i = 0; i < n; i++) {
            int idx = binarySearch(potions, 0, m, ((int) Math.ceil(success * 1.0 / spells[i])));
            result[i] = m - idx;
        }
        return result;
    }

    public int binarySearch(int[] potions, int left, int right, int target) {
        while (left < right) {
            int mid = left + (right - left) / 2;
            if (potions[mid] < target) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        return left;
    }

    public static void main(String[] args) {
        SuccessfulPairsOfSpellsAndPotionsSolution1 sol = new SuccessfulPairsOfSpellsAndPotionsSolution1();
        System.out.println(
                "Output: " + Arrays.toString(sol.successfulPairs(new int[] { 5, 1, 3 }, new int[] { 1, 2, 3, 4, 5 }, 7))
                        + ", expected: [4,0,3]");
        System.out.println(
                "Output: " + Arrays.toString(sol.successfulPairs(new int[] { 3, 1, 2 }, new int[] { 8, 5, 8 }, 16))
                        + ", expected: [2,0,2]");
    }
}

File: 2300-successful-pairs-of-spells-and-potions-1.py
import math

class Solution:
    def successfulPairs(self, spells: list[int], potions: list[int], success: int) -> list[int]:
        '''
        len (result) = len(spells)
        We can sort the potions in ascending order and do binary search for the index with a value that is equal or bigger than (`success` + 1 // spells[i])
        '''
        n, m = len(spells), len(potions)
        result = [0] * n
        potions.sort()

        def binary_search(left, right, target):
            while left < right:
                mid = left + (right - left) // 2
                if potions[mid] < target:
                    left = mid + 1
                else:
                    right = mid

            return left

        for i in range(n):
            spell = spells[i]
            idx = binary_search(0, m, math.ceil(success / spell))
            result[i] = max(0, m - idx)
            
        return result


sol = Solution()
spells = [5, 1, 3]
potions = [1, 2, 3, 4, 5]
success = 7
print(
    f'output: {sol.successfulPairs(spells, potions, success)}, expected: {[4, 0, 3]}')
spells = [3, 1, 2]
potions = [8, 5, 8]
success = 16
print(
    f'output: {sol.successfulPairs(spells, potions, success)}, expected: {[2, 0, 2]}')
